### PR TITLE
Jonmv/http url

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden or @bratseth please review.

### CODE REVIEW

File: container-core/src/main/java/com/yahoo/restapi/HttpURL.java [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+            return new Path<>(segments.subList(offset, segments.size()), trailingSlash, validator, inverse);
+        }
+
+        /** Returns a copy of this which eliminates the given number of segments, from the end up. */
+        public Path<T> headPath(int offset) {
+            return new Path<>(segments.subList(0, segments.size() - offset), trailingSlash, validator, inverse);
+        }
+
+        /** Returns a copy of this with the <em>decoded</em> segment appended at the end; it may not be either of {@code ""}, {@code "."} or {@code ".."}. */
+        public Path<T> with(T segment) {
```
Comment (bratseth) [inline container-core/src/main/java/com/yahoo/restapi/HttpURL.java:222]: Missing validation of segment here?

**Review by bratseth — COMMENTED** [review]
Great! I think we need lots more unit tests of this class to cover the API surface.

**Comment by bratseth** [comment]
And do we really need the type argument for path segments?

File: container-core/src/main/java/com/yahoo/restapi/HttpURL.java (hunk 1 above)
Comment (jonmv) [reply container-core/src/main/java/com/yahoo/restapi/HttpURL.java:222]: Thanks! Yes, lost it when refactoring.

**Comment by jonmv** [comment]
Isn't it nice that validating all segments is as easy as adding a constructor argument when creating the path? Also, all things you read out of it are already wrapped, if you want to send the wrapped strings further down. 
If you don't want to use it, you can simply ignore it too, as `<String>` is an option. 

Well, this was the main reason I wanted an early review! 

(And more unit tests are definitely needed.)

**Comment by bratseth** [comment]
I guess some usage examples in tests would help.

**Comment by jonmv** [comment]
The idea is that the `HttpURL` can be used both for parsing (and validating parsed content) of incoming `URI` objects, in handlers; and also be used to construct `URI`s for outgoing requests.

**Comment by bratseth** [comment]
And why do segments have to be of multiple implementation types to do that?

**Comment by jonmv** [comment]
Different APIs may have different levels of strictness? The /doc/v1 must accept path segments with `/` for example, but in general, we should not allow that.

**Comment by jonmv** [comment]
We can't find a one-size-fits-all validation for these segments, so I thought the best thing was to make it as easy as possible to define your own validation.

**Comment by bratseth** [comment]
I see. Wouldn't it be clearer to accept a validator then?

**Comment by jonmv** [comment]
Isn't that exactly what's done here? `HttpURL.parse(uri, MyStrictSegment::from)` does that. But you don't want it to be part of the class signature? You would prefer its `url.path().segments()` to always return a `List<String>`, rather than `List<MyStrictSegment>`? But why? Wouldn't you want to pass on the wrapper that indicates what kind of string you're dealing with here? Also, when using it to construct a new URL, isn't it nice that you can see what kind of strings it will accept, if you're getting the `HttpURL` passed form somewhere else?

**Comment by jonmv** [comment]
I think the best thing would be for API declarations to contain the validation for each segment, tbh., but that's more tricky. It's also not exclusive with this. This could be used to provide a baseline validation for "everything". We could also remove the `<String>` possibility, and have a more lax default, and then require an explicit `RawString` argument for allowing "any" string as segments/queries.

**Comment by jonmv** [comment]
I can use this now in one of our handlers, as an example, perhaps?

**Comment by bratseth** [comment]
> But you don't want it to be part of the class signature? You would prefer its url.path().segments() to always return a List<String>, rather than List<MyStrictSegment>?

Not prefer maybe, but it's what I would expect because I think of the path segments as short-lived and not worthy of having their own classes.

I you feel this is better then feel free to merge, but I think we should document it since I didn't get it.

**Comment by jonmv** [comment]
Right, I understand that expectation. I'll try using it, and see if the type parameter is a blessing, or just a curse.

**Pull request merged**